File: shared/src/types/quiz.ts
/**
 * Mirrors the assessment model in DATABASE_DESIGN.md §4. Note
 * `QuestionForAttempt` deliberately omits `isCorrect`/answer-key data —
 * DATABASE_SECURITY.md §5 requires the backend to strip that before
 * serving an in-progress quiz to a `user`-role client. The full
 * `is_correct` shape is intentionally not modeled here at all, to avoid
 * accidentally giving a client-facing type that includes it.
 */
export type QuestionType =
  | "multiple_choice"
  | "true_false"
  | "short_answer"
  | "fill"
  | "match"
  | "order"
  | "open";
export type QuizAttemptStatus = "in_progress" | "submitted" | "graded";

export interface Quiz {
  id: string;
  subjectId: string;
  lectureId: string | null;
  title: string;
  description: string | null;
  timeLimitSeconds: number | null;
  status: "draft" | "published";
}

export interface QuestionOptionForAttempt {
  id: string;
  optionText: string;
  orderIndex: number;
}

/** One item on either side of a `match` question, as delivered to the
 * learner. `id` is the underlying `question_pairs.id` — shared by the
 * left and right entry of the same authoritative pair, but this alone
 * does not reveal which `right` array *position* is the correct match
 * for a given `left` entry (Phase 12F-BE contract review §C). */
export interface MatchItemForAttempt {
  id: string;
  text: string;
}

/** One item of a `order` question, as delivered to the learner in
 * shuffled order. `correct_order_index` is never included (Phase 12F-BE
 * contract review §D). */
export interface OrderItemForAttempt {
  id: string;
  text: string;
}

export interface QuestionForAttempt {
  id: string;
  questionType: QuestionType;
  prompt: string;
  points: number;
  options: QuestionOptionForAttempt[] | null;
  /** Present only for `questionType === "match"`; `right` is shuffled
   * per-request server-side. Null for every other type. */
  matchItems: { left: MatchItemForAttempt[]; right: MatchItemForAttempt[] } | null;
  /** Present only for `questionType === "order"`; shuffled per-request
   * server-side. Null for every other type. */
  orderItems: OrderItemForAttempt[] | null;
}

export interface QuizAttempt {
  id: string;
  quizId: string;
  userId: string;
  status: QuizAttemptStatus;
  startedAt: string;
  submittedAt: string | null;
  score: number | null;
}

/** One submitted pairing for a `match` question: `leftId`/`rightId` are
 * both `question_pairs.id` values (see `MatchItemForAttempt`). */
export interface MatchAnswerPair {
  leftId: string;
  rightId: string;
}

export interface SubmitAnswerInput {
  questionId: string;
  selectedOptionId?: string;
  answerText?: string;
  /** `match` questions only — one entry per left item. */
  matchAnswer?: MatchAnswerPair[];
  /** `order` questions only — `question_items.id` values in the
   * learner's chosen order. */
  orderAnswer?: string[];
}

/** Acknowledges that an answer was recorded — deliberately carries no
 * correctness/scoring information (PHASE 09B "Answer Submission": the
 * client never learns whether an individual answer is right until the
 * attempt is submitted and graded, and even then only via the aggregate
 * `QuizAttemptResult`, never a per-question answer key). */
export interface SubmitAnswerAck {
  questionId: string;
  recorded: boolean;
}

/** One previously-recorded answer for an attempt, as returned to the
 * owning learner for resuming/re-hydrating their in-progress selections
 * (PHASE 09B "Quiz Navigation"). Deliberately carries only what the UI
 * needs to re-select an option or re-populate a text answer — no
 * `isCorrect`/points/any answer-key data. */
export interface AttemptAnswer {
  questionId: string;
  selectedOptionId: string | null;
  answerText: string | null;
  /** Re-hydration for a previously-submitted `match` answer, or null. */
  matchAnswer: MatchAnswerPair[] | null;
  /** Re-hydration for a previously-submitted `order` answer, or null. */
  orderAnswer: string[] | null;
}

/** Server-computed result of a submitted attempt. Never includes the
 * answer key or per-question correctness — only the aggregate the
 * approved requirements support (PHASE 09B "Result Security"). */
export interface QuizAttemptResult {
  attemptId: string;
  quizId: string;
  status: QuizAttemptStatus;
  totalQuestions: number;
  answeredQuestions: number;
  correctAnswers: number;
  score: number;
  percentage: number;
  submittedAt: string | null;
  /** True while `status === "submitted"` and at least one `open`
   * question's answer has not yet been manually reviewed. `score`/
   * `percentage` in that state exclude unreviewed `open` questions'
   * points (Phase 12F-BE contract review §F). Absent/false once
   * `status === "graded"`. */
  pendingManualReview: boolean;
}

File: mobile/lib/shared/models/quiz.dart
/// Dart mirrors of `shared/src/types/quiz.ts`. `QuestionForAttempt` (and
/// its options) deliberately have NO `isCorrect`/answer-key field
/// anywhere in this file, matching the TypeScript contract exactly —
/// PHASE 10 §13's central requirement. There is no code path in this
/// class that could hold that data even if the backend mistakenly sent
/// it (`fromJson` never reads such a key).
class Quiz {
  const Quiz({
    required this.id,
    required this.subjectId,
    required this.lectureId,
    required this.title,
    required this.description,
    required this.timeLimitSeconds,
    required this.status,
  });

  final String id;
  final String subjectId;
  final String? lectureId;
  final String title;
  final String? description;
  final int? timeLimitSeconds;
  final String status;

  factory Quiz.fromJson(Map<String, dynamic> json) {
    return Quiz(
      id: json['id'] as String,
      subjectId: json['subjectId'] as String,
      lectureId: json['lectureId'] as String?,
      title: json['title'] as String,
      description: json['description'] as String?,
      timeLimitSeconds: json['timeLimitSeconds'] as int?,
      status: json['status'] as String,
    );
  }
}

class QuestionOptionForAttempt {
  const QuestionOptionForAttempt({
    required this.id,
    required this.optionText,
    required this.orderIndex,
  });

  final String id;
  final String optionText;
  final int orderIndex;

  factory QuestionOptionForAttempt.fromJson(Map<String, dynamic> json) {
    return QuestionOptionForAttempt(
      id: json['id'] as String,
      optionText: json['optionText'] as String,
      orderIndex: json['orderIndex'] as int,
    );
  }
}

class QuestionForAttempt {
  const QuestionForAttempt({
    required this.id,
    required this.questionType,
    required this.prompt,
    required this.points,
    required this.options,
  });

  final String id;
  final String questionType; // 'multiple_choice' | 'true_false' | 'short_answer'
  final String prompt;
  final int points;
  final List<QuestionOptionForAttempt>? options;

  factory QuestionForAttempt.fromJson(Map<String, dynamic> json) {
    final optionsJson = json['options'] as List<dynamic>?;
    return QuestionForAttempt(
      id: json['id'] as String,
      questionType: json['questionType'] as String,
      prompt: json['prompt'] as String,
      points: json['points'] as int,
      options: optionsJson
          ?.map((o) => QuestionOptionForAttempt.fromJson(o as Map<String, dynamic>))
          .toList(),
    );
  }
}

class QuizAttempt {
  const QuizAttempt({
    required this.id,
    required this.quizId,
    required this.userId,
    required this.status,
    required this.startedAt,
    required this.submittedAt,
    required this.score,
  });

  final String id;
  final String quizId;
  final String userId;
  final String status; // 'in_progress' | 'submitted' | 'graded'
  final String startedAt;
  final String? submittedAt;
  final num? score;

  factory QuizAttempt.fromJson(Map<String, dynamic> json) {
    return QuizAttempt(
      id: json['id'] as String,
      quizId: json['quizId'] as String,
      userId: json['userId'] as String,
      status: json['status'] as String,
      startedAt: json['startedAt'] as String,
      submittedAt: json['submittedAt'] as String?,
      score: json['score'] as num?,
    );
  }
}

/// Ack for a recorded answer — no correctness/score field, matching
/// `SubmitAnswerAck` exactly (PHASE 10 §13).
class SubmitAnswerAck {
  const SubmitAnswerAck({required this.questionId, required this.recorded});

  final String questionId;
  final bool recorded;

  factory SubmitAnswerAck.fromJson(Map<String, dynamic> json) {
    return SubmitAnswerAck(
      questionId: json['questionId'] as String,
      recorded: json['recorded'] as bool,
    );
  }
}

/// Server-computed result — the aggregate only, never per-question
/// correctness or an answer key (mirrors `QuizAttemptResult`).
class QuizAttemptResult {
  const QuizAttemptResult({
    required this.attemptId,
    required this.quizId,
    required this.status,
    required this.totalQuestions,
    required this.answeredQuestions,
    required this.correctAnswers,
    required this.score,
    required this.percentage,
    required this.submittedAt,
  });

  final String attemptId;
  final String quizId;
  final String status;
  final int totalQuestions;
  final int answeredQuestions;
  final int correctAnswers;
  final num score;
  final num percentage;
  final String? submittedAt;

  factory QuizAttemptResult.fromJson(Map<String, dynamic> json) {
    return QuizAttemptResult(
      attemptId: json['attemptId'] as String,
      quizId: json['quizId'] as String,
      status: json['status'] as String,
      totalQuestions: json['totalQuestions'] as int,
      answeredQuestions: json['answeredQuestions'] as int,
      correctAnswers: json['correctAnswers'] as int,
      score: json['score'] as num,
      percentage: json['percentage'] as num,
      submittedAt: json['submittedAt'] as String?,
    );
  }
}
